### PR TITLE
Fix kv-cacheing and bsz > 1 in eval recipe

### LINKED ISSUE #1600: Fix KV-cacheing + bsz > 1 with eval recipe
Currently, the eleuther recipe won't actually work with bsz>1 and KV-cacheing because some batches recieved by `_model_generate` are smaller than the batch size specified. In the code, there's this comment:

```python
        # Setup caches for a given batch size
        # Technically this is not necessary, but it's a good way to ensure that
        # the caches won't error on a different batch size. In addition, caches
        # are not needed for a regular model call, so we just setup here
```
Which also alludes to this, but it makes repeated calls to `setup_caches` to no avail, because there's currently no way to completely tear down caches. `reset_caches` just zeros them out.

cc @joecummings

### PR DESCRIPTION
**Pull request by SalmanMohammadi**
#### Context
What is the purpose of this PR? Is it to
- [ ] add a new feature
- [x] fix a bug
- [ ] update tests and/or documentation
- [ ] other (please add here)

Please link to any issues this PR addresses.
closes #1600

#### Changelog
What are the changes made in this PR?
* The eval harness will return `batch_size` specified prompts until the last batch, where it will return a smaller number of samples.
* This breaks our static KV-cacheing which is only setup for a single batch size. 
* To address this, we can manually pad out batches, and then remove the padded samples after generation.
* This could be wasteful for large batch sizes, and where the last batch may be mostly empty. We can kind of mitigate this by padding out with `self._tokenizer.eos_id` so we correctly trigger early stopping in generation.
* Expected test values are changed because `limit` has changed.

* Since swapping between KV-cacheing-inference and regular inference isn't supported yet, I've guarded against using generation tasks with non-generation tasks. This will be addressed in a follow up.
* I've also moved setup cache logic outside of `_model_generate` - it only needs to be called once if we're expecting a generation task.

Thanks to @baberabb for the advice here.

#### Test plan
Please make sure to do each of the following if applicable to your PR. If you're unsure about any one of these just ask and we will happily help. We also have a [contributing page](https://github.com/pytorch/torchtune/blob/main/CONTRIBUTING.md) for some guidance on contributing.

- [x] run pre-commit hooks and linters (make sure you've first installed via `pre-commit install`)
- [x] add [unit tests](https://github.com/pytorch/torchtune/tree/main/tests/torchtune) for any new functionality
- [ ] update [docstrings](https://github.com/pytorch/torchtune/tree/main/docs/source) for any new or updated methods or classes
- [x] run unit tests via `pytest tests`
- [x] run recipe tests via `pytest tests -m integration_test`
- [ ] manually run any new or modified recipes with sufficient proof of correctness
- [ ] include relevant commands and any other artifacts in this summary (pastes of loss curves, eval results, etc.)

#### UX
If your function changed a public API, please add a dummy example of what the user experience will look like when calling it.
Here is a [docstring example](https://github.com/pytorch/torchtune/blob/6a7951f1cdd0b56a9746ef5935106989415f50e3/torchtune/modules/vision_transformer.py#L285)
and a [tutorial example](https://pytorch.org/torchtune/main/tutorials/qat_finetune.html#applying-qat-to-llama3-models)

- [ ] I did not change any public API
- [ ] I have added an example to docs or docstrings

### CODE REVIEW

File: recipes/eleuther_eval.py [hunk 1]
```diff
@@ -130,20 +123,11 @@ def _model_generate(
...
-        # the caches won't error on a different batch size. In addition, caches
-        # are not needed for a regular model call, so we just setup here
-        if self.enable_kv_cache:
-            with context.device:
-                self._model.setup_caches(batch_size=curr_batch_size, dtype=self._dtype)
+        maybe_padded_context = torch.nn.functional.pad(
+            context,
+            (0, 0, 0, self._batch_size - curr_batch_size),
+            value=self._tokenizer.eos_id,
+        )
```
Comment (ebsmothers) [inline recipes/eleuther_eval.py:130]: nit: add a comment explaining this cause otherwise it's not super obvious
